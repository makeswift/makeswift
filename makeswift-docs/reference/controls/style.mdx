---
title: Style
description: Add a Style control to your component in the Makeswift builder to visually edit a CSS class `string` prop.
icon: "droplet"
---

<Frame type="glass" caption="Style controls for an Image component">
  <img src="/images/style-control.png" />
</Frame>

The Style control is the only control that is responsive; its values can be changed at different viewport sizes without affecting the others. For example, you could set a width of 1200px in desktop and 100% in mobile.

## Params

<ParamField query="properties" type="StyleProperty[] | Style.All | Style.Default" default="Style.Default">
  An array of style properties that can be visually edited. These values include:

- `Style.Width`
- `Style.Margin`
- `Style.Padding`
- `Style.Border`
- `Style.BorderRadius`
- `Style.TextStyle`

The following presets are also available:

- `Style.Default` which is equivalent to `[Style.Width, Style.Margin]`
- `Style.All` which includes all properties.

</ParamField>

## Example

The following example adds a Style control to the `className` prop of an Image component.

<CodeGroup>

```tsx Image.makeswift.ts
import { Style } from "@makeswift/runtime/controls"

import { runtime } from "@/makeswift/runtime"

import { Image } from "./Image"

runtime.registerComponent(Image, {
  type: "image",
  label: "Image",
  props: {
    className: Style({ properties: Style.All }),
  },
})
```

```tsx Image.tsx
interface Props {
  className?: string
}

export function Image({ className }: Props) {
  return <img className={className} src="https://via.placeholder.com/150" />
}
```

</CodeGroup>

import NamingConventions from "/snippets/naming-conventions.mdx"

<NamingConventions />
