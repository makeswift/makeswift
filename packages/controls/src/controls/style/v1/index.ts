export * as StyleSchema from './schema'
export * from './types'
export * from './style'
export * from './style-control'
