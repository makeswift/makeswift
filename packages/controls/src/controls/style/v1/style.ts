import { match, P } from 'ts-pattern'
import { z } from 'zod'

import { mapValues, nullToUndefined } from '../../../lib/functional'
import { StableValue } from '../../../lib/stable-value'
import { safeParse, type ParseResult } from '../../../lib/zod'

import { type ResponsiveValue } from '../../../common'
import { responsiveValue } from '../../../common/schema'
import { type CopyContext } from '../../../context'
import { Targets, type IntrospectionTarget } from '../../../introspection'
import {
  ResourceSchema,
  type ColorData,
  type ResolvedColorData,
} from '../../../resources'
import { type ResourceResolver } from '../../../resources/resolver'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../../serialization'
import { type Stylesheet } from '../../../stylesheet'
import { Color } from '../../color'
import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../../definition'
import { type SendMessage } from '../../instance'

import * as StyleSchema from './schema'
import { StyleControl } from './style-control'
import { type BorderData, type BorderSideData } from './types'

type Config = z.infer<typeof Definition.schema.config>

type Schema<_C extends Config> = typeof Definition.schema
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type InstanceType<_C extends Config> = StyleControl

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>,
  InstanceType<C>
> {
  static readonly type = 'makeswift::controls::style' as const
  static readonly Property = {
    Width: `${this.type}::property::width`,
    Margin: `${this.type}::property::margin`,
    Padding: `${this.type}::property::padding`,
    Border: `${this.type}::property::border`,
    BorderRadius: `${this.type}::property::border-radius`,
    TextStyle: `${this.type}::property::text-style`,
  } as const

  static readonly DefaultProperties: StyleProperty[] = [
    this.Property.Width,
    this.Property.Margin,
  ]

  static readonly AllProperties: StyleProperty[] = Object.values(this.Property)

  static get schema() {
    const type = z.literal(Definition.type)

    const properties = z.union([
      z.literal(this.Property.Width),
      z.literal(this.Property.Margin),
      z.literal(this.Property.Padding),
      z.literal(this.Property.Border),
      z.literal(this.Property.BorderRadius),
      z.literal(this.Property.TextStyle),
    ])

    const config = z.object({
      properties: z.array(properties),
    })

    const dataSchema = <C extends z.ZodTypeAny>(color: C) =>
      z.object({
        width: responsiveValue(StyleSchema.width).optional(),
        margin: responsiveValue(StyleSchema.margin).optional(),
        padding: responsiveValue(StyleSchema.padding).optional(),
        border: responsiveValue(StyleSchema.borderSchema(color)).optional(),
        borderRadius: responsiveValue(StyleSchema.borderRadius).optional(),
        textStyle: responsiveValue(StyleSchema.textStyle).optional(),
      })

    const data = dataSchema(ResourceSchema.colorData)
    const resolvedData = dataSchema(ResourceSchema.resolvedColorData)

    const value = data
    const resolvedValue = z.string()

    const definition = z.object({
      type,
      config,
    })

    return {
      type,
      config,
      definition,
      data,
      resolvedData,
      value,
      resolvedValue,
    }
  }

  static deserialize(data: DeserializedRecord): StyleDefinition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Style: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    const { config } = Definition.schema.definition.parse(data)
    return Style(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return {
      ...Definition.schema,
      data: Definition.schema.data as SchemaType<DataType<C>>,
      value: Definition.schema.value as SchemaType<ValueType<C>>,
    }
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    return this.schema.data.optional().parse(data)
  }

  toData(value: ValueType<C>): DataType<C> {
    return value
  }

  copyData(
    data: DataType<C> | undefined,
    context: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    return {
      ...data,
      border: mapBorderSides(data.border, (side) =>
        side.color == null
          ? side
          : {
              ...side,
              color: Color().copyData(side.color, context),
            },
      ),
    }
  }

  resolveValue(
    data: DataType<C> | undefined,
    resolver: ResourceResolver,
    stylesheet: Stylesheet,
    control?: InstanceType<C>,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    const resolvedColors = new Map<
      string,
      Resolvable<ResolvedColorData | undefined>
    >(
      [...borderColors(data?.border)].map((color) => [
        color.swatchId,
        Color().resolveSwatch(nullToUndefined(color), resolver),
      ]),
    )

    const resolvedColor = (swatchId: string | undefined) =>
      swatchId ? resolvedColors.get(swatchId)?.readStable() : undefined

    const resolvedStyle = (data: DataType<C>) => ({
      ...data,
      border: mapBorderSides(data.border, (side) => ({
        ...side,
        color: resolvedColor(side?.color?.swatchId),
      })),
    })

    const stableValue = StableValue({
      name: Definition.type,
      read: () =>
        stylesheet.defineStyle(
          {
            properties: this.config.properties,
            styleData: data != null ? resolvedStyle(data) : {},
          },
          (boxModel) => control?.changeBoxModel(boxModel),
        ),
      deps: [...resolvedColors.values()],
    })

    return {
      ...stableValue,
      triggerResolve: () =>
        Promise.all(
          [...resolvedColors.values()].map((r) => r.triggerResolve()),
        ),
    }
  }

  createInstance(sendMessage: SendMessage) {
    return new StyleControl(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  introspect<R>(data: DataType<C> | undefined, target: IntrospectionTarget<R>) {
    if (data == null || target.type !== Targets.Swatch.type) {
      return []
    }

    return Object.values(data)
      .flat()
      .flatMap((override) =>
        match(override.value)
          .with(
            P.union(
              { borderTop: { color: P.any } },
              { borderLeft: { color: P.any } },
              { borderBottom: { color: P.any } },
              { borderRight: { color: P.any } },
            ),
            (overrideValue) => [
              overrideValue.borderBottom?.color,
              overrideValue.borderTop?.color,
              overrideValue.borderLeft?.color,
              overrideValue.borderRight?.color,
            ],
          )
          .otherwise(() => [null]),
      )
      .flatMap((color) => target.introspect(color))
  }
}

function* borderColors(
  responsibeBorder: ResponsiveValue<BorderData> | undefined,
): Generator<ColorData> {
  for (const deviceBorder of responsibeBorder ?? []) {
    const sides = deviceBorder.value
    for (const side of Object.values(sides)) {
      if (side?.color != null) {
        yield side?.color
      }
    }
  }
}

function mapBorderSides<R>(
  responsibeBorder: ResponsiveValue<BorderData> | undefined,
  callback: (side: BorderSideData) => R,
) {
  return responsibeBorder?.map((deviceBorder) => ({
    ...deviceBorder,
    value: mapValues(
      deviceBorder.value,
      (side: BorderSideData | undefined | null) =>
        side == null ? side : callback(side),
    ),
  }))
}

export class StyleDefinition<C extends Config = Config> extends Definition<C> {}

export type StyleProperty =
  (typeof StyleDefinition.Property)[keyof typeof StyleDefinition.Property]

export type ResolvedStyleData = z.infer<
  StyleDefinition['schema']['resolvedData']
>

export function Style(config?: Config): StyleDefinition<Config> {
  return new StyleDefinition<Config>(
    config ?? {
      properties: StyleDefinition.DefaultProperties,
    },
  )
}

Style.Default = StyleDefinition.DefaultProperties
Style.All = StyleDefinition.AllProperties
Style.Width = StyleDefinition.Property.Width
Style.Margin = StyleDefinition.Property.Margin
Style.Padding = StyleDefinition.Property.Padding
Style.Border = StyleDefinition.Property.Border
Style.BorderRadius = StyleDefinition.Property.BorderRadius
Style.TextStyle = StyleDefinition.Property.TextStyle
