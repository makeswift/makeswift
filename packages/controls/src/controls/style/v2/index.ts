export * from './style-v2'
export * from './style-v2-control'
