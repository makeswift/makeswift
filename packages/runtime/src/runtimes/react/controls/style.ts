import { CSSObject } from '@emotion/serialize'
import {
  Style,
  type StyleProperty,
  type ResolvedBorderSideData,
  type FontSizePropertyData,
  type WidthPropertyData,
  type ResolvedStyleData,
} from '@makeswift/controls'

import { colorToString } from '../../../components/utils/colorToString'
import { responsiveStyle } from '../../../components/utils/responsive-style'

import { BorderRadiusLonghandPropertyData } from '../../../css/border-radius'
import { lengthPercentageDataToString } from '../../../css/length-percentage'
import { marginPropertyDataToStyle } from '../../../css/margin'
import { paddingPropertyDataToStyle } from '../../../css/padding'
import { Breakpoints } from '../../../state/modules/breakpoints'

const defaultMargin = {
  marginTop: 0,
  marginRight: 'auto',
  marginBottom: 0,
  marginLeft: 'auto',
}

const defaultPadding = {
  paddingTop: 0,
  paddingRight: 0,
  paddingBottom: 0,
  paddingLeft: 0,
}

export function styleV1Css(
  breakpoints: Breakpoints,
  style: ResolvedStyleData | undefined,
  properties: StyleProperty[],
): CSSObject {
  return {
    ...(properties.includes(Style.Width) && {
      maxWidth: '100%',
    }),
    ...responsiveStyle(
      breakpoints,
      [
        style?.width,
        style?.margin,
        style?.padding,
        style?.border,
        style?.borderRadius,
        style?.textStyle,
      ] as const,
      ([width, margin, padding, border, borderRadius, textStyle]) => ({
        ...(properties.includes(Style.Width) && {
          width: widthToString(width) ?? '100%',
        }),
        ...(properties.includes(Style.Margin) &&
          marginPropertyDataToStyle(margin ?? defaultMargin, defaultMargin)),
        ...(properties.includes(Style.Padding) &&
          paddingPropertyDataToStyle(padding ?? defaultPadding, defaultPadding)),
        ...(properties.includes(Style.Border) && {
          borderTop: borderSideToString(border?.borderTop) ?? '0 solid black',
          borderRight: borderSideToString(border?.borderRight) ?? '0 solid black',
          borderBottom: borderSideToString(border?.borderBottom) ?? '0 solid black',
          borderLeft: borderSideToString(border?.borderLeft) ?? '0 solid black',
        }),
        ...(properties.includes(Style.BorderRadius) && {
          borderTopLeftRadius: borderRadiusToString(borderRadius?.borderTopLeftRadius) ?? 0,
          borderTopRightRadius: borderRadiusToString(borderRadius?.borderTopRightRadius) ?? 0,
          borderBottomRightRadius: borderRadiusToString(borderRadius?.borderBottomRightRadius) ?? 0,
          borderBottomLeftRadius: borderRadiusToString(borderRadius?.borderBottomLeftRadius) ?? 0,
        }),
        ...(properties.includes(Style.TextStyle) && {
          ...(textStyle?.fontFamily && { fontFamily: `"${textStyle.fontFamily}"` }),
          ...(textStyle?.letterSpacing && { letterSpacing: textStyle.letterSpacing }),
          ...(textStyle?.fontSize && { fontSize: fontSizeToString(textStyle.fontSize) }),
          ...(textStyle?.fontWeight && { fontWeight: textStyle.fontWeight }),
          textTransform: textStyle?.textTransform ?? [],
          fontStyle: textStyle?.fontStyle ?? [],
        }),
      }),
    ),
  }

  function widthToString(widthProperty: WidthPropertyData | undefined): string | null {
    if (widthProperty == null) return null

    return lengthPercentageDataToString(widthProperty)
  }

  function borderSideToString(
    borderSide: ResolvedBorderSideData | null | undefined,
  ): string | null {
    if (borderSide == null) return null

    const { width, color, style } = borderSide
    return `${width != null ? width : 0}px ${style} ${
      color != null ? colorToString(color) : 'black'
    }`
  }

  function borderRadiusToString(
    borderRadius: BorderRadiusLonghandPropertyData | null | undefined,
  ): string | null {
    if (borderRadius == null) return null

    return lengthPercentageDataToString(borderRadius)
  }

  function fontSizeToString(fontSize: NonNullable<FontSizePropertyData>) {
    return `${fontSize.value}${fontSize.unit}`
  }
}
