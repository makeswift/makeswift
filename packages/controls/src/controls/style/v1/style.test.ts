import { Style, StyleDefinition } from './style'

describe('Style', () => {
  describe('constructor', () => {
    test('returns correct definition', () => {
      expect(Style()).toMatchSnapshot('default')
      expect(
        Style({ properties: [Style.Border, Style.Margin] }),
      ).toMatchSnapshot('border and margin')

      expect(Style({ properties: Style.All })).toMatchSnapshot('all')
    })

    test('disallows extraneous properties', () => {
      Style({
        properties: [Style.Border],
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: StyleDefinition) {}
    assignTest(Style())
    assignTest(Style({ properties: [Style.Border] }))
    assignTest(Style({ properties: [Style.Border, Style.Padding] }))
    assignTest(Style({ properties: Style.All }))
  })
})
