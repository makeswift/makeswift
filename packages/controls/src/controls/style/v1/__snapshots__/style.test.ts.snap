// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Style constructor returns correct definition: all 1`] = `
StyleDefinition {
  "config": {
    "properties": [
      "makeswift::controls::style::property::width",
      "makeswift::controls::style::property::margin",
      "makeswift::controls::style::property::padding",
      "makeswift::controls::style::property::border",
      "makeswift::controls::style::property::border-radius",
      "makeswift::controls::style::property::text-style",
    ],
  },
}
`;

exports[`Style constructor returns correct definition: border and margin 1`] = `
StyleDefinition {
  "config": {
    "properties": [
      "makeswift::controls::style::property::border",
      "makeswift::controls::style::property::margin",
    ],
  },
}
`;

exports[`Style constructor returns correct definition: default 1`] = `
StyleDefinition {
  "config": {
    "properties": [
      "makeswift::controls::style::property::width",
      "makeswift::controls::style::property::margin",
    ],
  },
}
`;
